README
The Nashville ColdFusion User Group publishes our Web site code
free of charge, for anyone to use or reference. All code is governed
by Apache License 2.0 (http://www.apache.org/licenses/LICENSE-2.0).

http://www.ncfug.com

Our Web site is based on the open source ColdFusion User Group Manager
application, nicknamed CHUG. This MVC application makes use of Mach-ii
(an MVC ColdFusion framework) and ColdSpring (an Inversion of Control
or IoC ColdFusion framework).

http://code.google.com/p/chug/

MachII/CHANGELOG.txt
------------------------------------------------
    __  ___           __          ________
   /  |/  /___ ______/ /_        /  _/  _/
  / /|_/ / __ `/ ___/ __ \______ / / / /  
 / /  / / /_/ / /__/ / / /_____// /_/ /   
/_/  /_/\__,_/\___/_/ /_/     /___/___/  

Version: 1.6.0.2
Built-on: 2008-09-22 23:00:49
$Id: CHANGELOG 891 2008-07-24 20:42:35Z peterfarrell $
------------------------------------------------
If version number and built-on are placeholders (e.g. @xyz@), you are 
using a  bleeding edge release (BER) of the framework.
------------------------------------------------

Copyright 2008 GreatBizTools, LLC

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

------------------------------------------------
 _                         
/ |_  _.._  _  _   |  _  _ 
\_| |(_|| |(_|(/_  |_(_)(_|
            _|           _|
------------------------------------------------
Fixed Defects
------------------------------------------------
* Fixed Inconsistent Cleanup of Abandoned Redirect Persist Data
Re-built cleanup routine to ensure total cleanup of abandoned redirect persist data.
(2007-10-20 pfarrell)

* Fixed SES / Friendly URLs with zero length param value breaks parseSES()
CF8 strips out the double "//" from the path info. This causes url params to become mismatched.
(2007-11-13 pfarrell)

* Fixed event arguments not available in handleException plugin point or exception event
The missing event is placed in an event arg named 'missingEvent' in the exception event. Replumbed 
when the exception event is placed in the queue, so getNextEvent() is now available in the 
handleException plugin point.

In the handleException plugin point, you can get the missing event by doing this:
arguments.eventContext.getNextEvent().getArg("missingEvent")

You use getNextEvent() because the handleException() plugin point is called before the 
next event begins processing.

In your exception event handler:
event.getArg("missingEvent")

If you clear the event queue in the handleException() plugin point and do not announce a
a new event.  You will clear the exception event and no exception event handler will be
run.  You must announce a new event if you clear the event queue in the handleException()
plugin point.

See for more information: http://greatbiztoolsllc-trac.cvsdude.com/mach-ii/ticket/14
(2007-11-21 pfarrell)

* Fixed EventContext.cfc failed to handle exception when an event-mapping exists
If an exception occurs in an event which has an event-mapping defined, the handleException
method of EventContext.cfc itself throws a new exception.

See for more information: http://greatbiztoolsllc-trac.cvsdude.com/mach-ii/ticket/24
(2008-01-14 pfarrell)

* AppKey with invalid characters for struct keys caused redirect persists to fail
The boostrappers now "clean" AppKeys so StructGet() built-in function does not
fail.

See for more information: http://greatbiztoolsllc-trac.cvsdude.com/mach-ii/ticket/39
(2008-06-15 pfarrell)

* Redirect caused exception when statusType is set 'permanent' or 'prg'
The redirect would occur, but an exception would occur causing the handleException()
plugin point to be run.

See for more information: http://greatbiztoolsllc-trac.cvsdude.com/mach-ii/ticket/43
(2008-06-20 pfarrell)

* Plugins in modules failed to run before parent plugins if runParent="before"
The PluginManager only set the runParent attribute when the XML was being parsed
as override XML.

See for more information: http://greatbiztoolsllc-trac.cvsdude.com/mach-ii/ticket/50
(2008-06-20 pfarrell)

------------------------------------------------
Improvements / Enhancements
------------------------------------------------
* Application.cfc Bootstrapper Enhancements and OnSessionStart / OnSessionEnd Integration
Add onSessionStart / onSessionEnd integration as new plugin point.  This enhancement helps 
developers to integrate third-party modules into a base application with ease and/or to 
encapsulate session application events into Mach-II rather than maintaining application 
event code in their Application.cfc file.

The bootstrapper (mach-ii.cfc) was added onApplicationStart and onRequestStart methods to
allow for easier setup of Mach-II applications. These methods can be overrided for custom
functionality.

See for more information: http://greatbiztoolsllc-trac.cvsdude.com/mach-ii/ticket/5
(2007-11-10 pfarrell)

* Message Subscriber Listeners Enhancement (AKA Broadcast Style Listeners Invocation)
This enhancement allows developers another way to invoke listeners by publishing messages.
This lets developers easily swap different listener methods or add addition listeners that 
respond to a message by changing the subscribes directive instead of changing the a bunch 
of notify commands throughout an XML file since listeners are decoupled from the event-handlers.

See for more information: http://greatbiztoolsllc-trac.cvsdude.com/mach-ii/ticket/4
(2007-11-11 pfarrell)

* Caching Enhancement
Added a generic interface that allows developers to implement their own caching strategy 
for caching at the controller level. Developers may specify which caching strategy they 
want to use through a special Mach-II caching property (MachII.caching.CachingProperty).
Additional configuration parameters may also be passed in to the caching strategy through 
the caching property. The default Mach-II caching strategy will leverage storing cached 
items in the application scope using the TimeSpanCache (MachII.caching.strategies.TimeSpanCache).


js/common.js
function dropDownNav()
{
    var sfEls = document.getElementById("nav").getElementsByTagName("LI");
    for (var i=0; i<sfEls.length; i++)
    {
        sfEls[i].onmouseover=function()
        {
            this.className += ' sfhover';
            this.style.zIndex=200;
        }
        sfEls[i].onmouseout=function()
        {
            this.className = this.className.replace(/\s?sfhover/g,'');
        }
    }
}

function windowLoad() { 
 	if (!document.getElementsByTagName) return; 
 	var anchors = document.getElementsByTagName("a"); 
 	for (var i=0; i<anchors.length; i++) { 
   		var anchor = anchors[i]; 
   		if (anchor.getAttribute("href") && anchor.getAttribute("rel") == "external") 
     		anchor.target = "_blank"; 
 	}
 	
 	if (document.getElementById('nav') != null)
    {
        dropDownNav();
    }
}

if (window.attachEvent)
{
    window.attachEvent("onload", windowLoad);
}

trunk/js/lib/qforms/wddx.js
/******************************************************************************
 qForm JSAPI: WDDX Mod Library

 Author: Dan G. Switzer, II
 Build:  101
******************************************************************************/

/******************************************************************************
 Required Functions
******************************************************************************/
function __serializeStruct(struct){
	// open packet
	var aWDDX = new Array("<wddxPacket version='1.0'><header/><data><struct>");
	for( var key in struct ) aWDDX[aWDDX.length] = "<var name='" + key.toLowerCase() + "'><string>" + __wddxValue(struct[key]) + "</string></var>";
	// close packet
	aWDDX[aWDDX.length] = "</struct></data></wddxPacket>";
	return aWDDX.join("");
}


function __wddxValue(str){
	var aValue = new Array();
	for( var i=0; i < str.length; ++i) aValue[aValue.length] = _encoding.table[str.charAt(i)];
	return aValue.join("");
}

function _wddx_Encoding(){
	// Encoding table for strings (CDATA)
	var et = new Array();

	// numbers to characters table
	var n2c = new Array();

	for( var i=0; i < 256; ++i ){
		// build a character from octal code
		var d1 = Math.floor(i/64);
		var d2 = Math.floor((i%64)/8);
		var d3 = i%8;
 		var c = eval("\"\\" + d1.toString(10) + d2.toString(10) + d3.toString(10) + "\"");

		// modify character-code conversion tables
		n2c[i] = c;

		// modify encoding table
		if( i < 32 && i != 9 && i != 10 && i != 13 ){
			// control characters that are not tabs, newlines, and carriage returns

			// create a two-character hex code representation
			var hex = i.toString(16);
			if( hex.length == 1 ) hex = "0" + hex;

			et[n2c[i]] = "<char code='" + hex + "'/>";

		} else if( i < 128 ){
			// low characters that are not special control characters
			et[n2c[i]] = n2c[i];
		} else {
			// high characters
			et[n2c[i]] = "&#x" + i.toString(16) + ";";
		}
	}

	// special escapes for CDATA encoding
	et["<"] = "&lt;";
	et[">"] = "&gt;";
	et["&"] = "&amp;";

	this.table = et;
}
_encoding = new _wddx_Encoding();


/******************************************************************************
 qForm Methods
******************************************************************************/
function _a_serialize(exclude){
	// if you need to reset the default values of the fields
	var lstExclude = (arguments.length > 0) ? "," + _removeSpaces(arguments[0]) + "," : "";
	var struct = new Object(), stcAllFields = qFormAPI.getFields();
	// loop through form elements
	for( var key in stcAllFields ){
		if( lstExclude.indexOf("," + key + ",") == -1 ) struct[key] = stcAllFields[key];
	}
	// create & return the serialized object
	return __serializeStruct(struct);
}
_a.prototype.serialize = _a_serialize;

// define qForm serialize(); prototype
function _q_serialize(exclude){
	// if you need to reset the default values of the fields
	var lstExclude = (arguments.length > 0) ? "," + _removeSpaces(arguments[0]) + "," : "";
	var struct = new Object();
	// loop through form elements
	for( var j=0; j < this._fields.length; j++ ){
		if( lstExclude.indexOf("," + this._fields[j] + ",") == -1 ) struct[this._fields[j]] = this[this._fields[j]].getValue();
	}
	// create & return the serialized object
	return __serializeStruct(struct);
}
qForm.prototype.serialize = _q_serialize;


trunk/js/lib/qforms/validation.js
/******************************************************************************
 qForm JSAPI: Validation Library

 Author: Dan G. Switzer, II
 Build:  113
******************************************************************************/
qFormAPI.packages.validation = true;

// define Field isNotNull(); prototype
function _Field_isNotNull(){
	// check for blank field
	if( this.value.length == 0 ){
		this.error = "You must specify a(n) " + this.description + ".";
	}
}
_addValidator("isNotNull", _Field_isNotNull, true);

// define Field isNotEmpty(); prototype
function _Field_isNotEmpty(){
	// check for blank field
	if( _ltrim(this.value).length == 0 ){
		this.error = "You must specify a(n) " + this.description + ".";
	}
}
_addValidator("isNotEmpty", _Field_isNotEmpty);

// define Field isEmail(); prototype
function _Field_isEmail(){
	// check for @ . or blank field
	if( this.value.indexOf(" ") != -1 ){
		this.error = "Invalid " + this.description + " address. An e-mail address should not contain a space.";
	} else if( this.value.indexOf("@") == -1 ){
		this.error = "Invalid " + this.description + " address. An e-mail address must contain the @ symbol.";
	} else if( this.value.indexOf("@") == 0 ){
		this.error = "Invalid " + this.description + " address. The @ symbol can not be the first character of an e-mail address.";
	} else if( this.value.substring(this.value.indexOf("@")+2).indexOf(".") == -1 ){
		this.error = "Invalid " + this.description + " address. An e-mail address must contain at least one period after the @ symbol.";
	} else if( this.value.lastIndexOf("@") == this.value.length-1 ){
		this.error = "Invalid " + this.description + " address. The @ symbol can not be the last character of an e-mail address.";
	} else if( this.value.lastIndexOf(".") == this.value.length-1 ){
		this.error = "Invalid " + this.description + " address. A period can not be the last character of an e-mail address.";
	}
}
_addValidator("isEmail", _Field_isEmail);

// define Field isPassword(); prototype
function _Field_isPassword(field, minlen, maxlen){
	var minlen = _param(arguments[1], 1, "number");   // default minimum length of password
	var maxlen = _param(arguments[2], 255, "number"); // default maximum length of password

	if( field != null && !this.compare(field) ){
		this.error = "The " + this.description + " and " + this.qForm[field].description + " values do not match.";
	}

	// if there aren't an errors yet
	if( this.error.length == 0 ){
		if( (this.value.length < minlen) || (this.value.length > maxlen) ){
			this.error = "The " + this.description + " field must be between " + minlen.toString() + " and " + maxlen.toString() + " characters long.";
		}
	}
}
_addValidator("isPassword", _Field_isPassword);

// define Field isSame(); prototype
function _Field_isSame(field){
	if( !this.compare(field) ){
		this.error = "The " + this.description + " and " + this.qForm[field].description + " values do not match.";
	}
}
_addValidator("isSame", _Field_isSame);

// define Field isDifferent(); prototype
function _Field_isDifferent(field){
	if( this.compare(field) ){
		this.error = "The " + this.description + " and " + this.qForm[field].description + " must be different.";
	}
}
_addValidator("isDifferent", _Field_isDifferent);

// define Field isRange(); prototype
function _Field_isRange(low, high){
	var low = _param(arguments[0], 0, "number");
	var high = _param(arguments[1], 9999999, "number");
	var iValue = parseInt(this.value, 10);
	if( isNaN(iValue) ) iValue = 0;

	// check to make sure the number is within the valid range
	if( ( low > iValue ) || ( high < iValue ) ){
		this.error = "The " + this.description + " field does not contain a\nvalue between " + low + " and " + high + ".";
	}
}
_addValidator("isRange", _Field_isRange);

// define Field isInteger(); prototype
function _Field_isInteger(){
	var i = parseInt(this.value, 10);
	// make sure the user specified a numeric value
	if( isNaN(i) || (String(i) != this.value) ){
		this.error = "The value for " + this.description + " is not a numeric value. This field requires a numeric value.";
	}
}
_addValidator("isInteger", _Field_isInteger);

// define Field isNumeric(); prototype
function _Field_isNumeric(){
	var i = parseFloat(this.value, 10);
	// make sure the user specified a numeric value
	if( isNaN(i) || (String(i) != this.value) ){
		this.error = "The value for " + this.description + " is not a numeric value. This field requires a numeric value.";
	}
}
_addValidator("isNumeric", _Field_isNumeric);

// define Field isAlpha(); prototype
function _Field_isAlpha(){
	if( !_isLength(this.value, this.value.length, "alpha") ){
		this.error = "The value for " + this.description + " must contain only alpha characters.";
	}
}
_addValidator("isAlpha", _Field_isAlpha);

// define Field isAlphaNumeric(); prototype
function _Field_isAlphaNumeric(){
	if( !_isLength(this.value, this.value.length, "alphanumeric") ){
		this.error = "The value for " + this.description + " must contain only alpha-numeric characters.";
	}
}
_addValidator("isAlphaNumeric", _Field_isAlphaNumeric);

// define Field isDate(); prototype
function _Field_isDate(mask){
	var strMask = _param(arguments[0], "mm/dd/yyyy");
	var iMaskMonth = strMask.lastIndexOf("m") - strMask.indexOf("m") + 1;
	var iMaskDay = strMask.lastIndexOf("d") - strMask.indexOf("d") + 1;
	var iMaskYear = strMask.lastIndexOf("y") - strMask.indexOf("y") + 1;

	var strDate = this.value;

	// find the delimiter
	var delim = "", lstMask = "mdy";
	for( var i=0; i < strMask.length; i++ ){
		if (lstMask.indexOf(strMask.substring(i, i+1)) == -1){
			delim = strMask.substring(i, i+1);
			break;
		}
  }
	aMask = strMask.split(delim);
	if( aMask.length == 3 ){
		dt = this.value.split(delim);
		if( dt.length != 3 ) this.error = "An invalid date was provided for " + this.description + " field.";
		for( i=0; i < aMask.length; i++ ){
			if( aMask[i].indexOf("m") > -1 ) var sMonth = dt[i];
			else if( aMask[i].indexOf("d") > -1 ) var sDay = dt[i];
			else if( aMask[i].indexOf("y") > -1 ) var sYear = dt[i];
		}
	} else if( mask.length == 1 ){
		var sMonth = this.value.substring(strMask.indexOf("m")-1, strMask.lastIndexOf("m"));
		var sDay = this.value.substring(strMask.indexOf("d")-1, strMask.lastIndexOf("d"));
		var sYear = this.value.substring(strMask.indexOf("y")-1, strMask.lastIndexOf("y"));
	} else {
		this.error = "An invalid date mask was provided for " + this.description + " field.";
	}

	var iMonth = parseInt(sMonth, 10);
	var iDay = parseInt(sDay, 10);
	var iYear = parseInt(sYear, 10);

	if( isNaN(iMonth) || sMonth.length > iMaskMonth ) iMonth = 0;
	if( isNaN(iDay) || sDay.length > iMaskDay ) iDay = 0;
	if( isNaN(sYear) || sYear.length != iMaskYear ) sYear = null;

	lst30dayMonths = ",4,6,9,11,";

	// if there are no other errors, check for input errors
	if( this.error.length == 0 ){
		if( sYear == null ){
			this.error = "An invalid year was provided for the " + this.description + " field. The year \n   should be a " + iMaskYear + " digit number.";
		} else if(  (iMonth < 1) || (iMonth > 12 ) ){
			this.error = "An invalid month was provided for " + this.description + " field.";
		} else {
			if( iYear < 100 ) var iYear = iYear + ((iYear > 20) ? 1900 : 2000);
			var iYear = (sYear.length == 4) ? parseInt(sYear, 10) : parseInt("20" + sYear, 10);
			if( lst30dayMonths.indexOf("," + iMonth + ",") > -1 ){
				if( (iDay < 1) || (iDay > 30 ) ) this.error = "An invalid day was provided for the " + this.description + " field.";
			} else if( iMonth == 2 ){
				if( (iDay < 1) || (iDay > 28 && !( (iDay == 29) && (iYear%4 == 0 ) ) ) ) this.error = "An invalid day was provided for the " + this.description + " field.";
			} else {
				if( (iDay < 1) || (iDay > 31 ) ) this.error = "An invalid day was provided for the " + this.description + " field.";
			}
		}
	}

}
_addValidator("isDate", _Field_isDate);


// define Field isCreditCard(); prototype
function _Field_isCreditCard(){
	var strCC = _stripInvalidChars(this.value, "numeric").toString();
	var isNumeric = (strCC.length > 0) ? true : false;

	if( isNumeric ){
		// now check mod10
		var dd = (strCC.length % 2 == 1) ? false : true;
		var cd = 0;
		var td;

		for( var i=0; i < strCC.length; i++ ){
			td = parseInt(strCC.charAt(i), 10);
			if( dd ){
				td *= 2;
				cd += (td % 10);
				if ((td / 10) >= 1.0) cd++;
				dd = false;
			} else {
				cd += td;
				dd = true;
			}
		}
		if( (cd % 10) != 0 ) this.error = "The credit card number entered in the " + this.description + " field is invalid.";
	} else {
		this.error = "The credit card number entered in the " + this.description + " field is invalid.";
	}
}
_addValidator("isCreditCard", _Field_isCreditCard);

// define Field isPhoneNumber(); prototype
function _Field_isPhoneNumber(len){
	var len = parseInt(_param(arguments[0], 10, "number"), 10);
	var description = (this.description == this.name.toLowerCase()) ? "phone number" : this.description;

	// check to make sure the phone is the correct length
	if( !_isLength(this.value, len) ){
		this.error = "The " + description + " field must include " + len + " digits.";
	}
}
_addValidator("isPhoneNumber", _Field_isPhoneNumber);

// define Field isLength(); prototype
function _Field_isLength(len, type){
	var len = parseInt(_param(arguments[0], 10, "number"), 10);
	var type = _param(arguments[1], "numeric");

	// check to make sure the phone is the correct length
	if( !_isLength(this.value, len, type) ){
		this.error = "The " + this.description + " field must include " + len + " " + type + " characters.";
	}
}
_addValidator("isLength", _Field_isLength);

// define Field isSSN(); prototype
function _Field_isSSN(){
	var description = (this.description == this.name.toLowerCase()) ? "social security" : this.description;

	// check to make sure the phone is the correct length
	if( !_isLength(this.value, 9) ){
		this.error = "The " + description + " field must include 9 digits.";
	}
}
_addValidator("isSSN", _Field_isSSN);


// define Field isState(); prototype
function _Field_isState(){
	// check to make sure the phone is the correct length
	if( _getState(this.value) == null ){
		this.error = "The " + this.description + " field must contain a valid 2-digit state abbreviation.";
	}
}
_addValidator("isState", _Field_isState);

// define Field isZipCode(); prototype
function _Field_isZipCode(){
	var description = (this.description == this.name.toLowerCase()) ? "zip code" : this.description;

	iZipLen = _stripInvalidChars(this.value).length;

	// check to make sure the zip code is the correct length
	if( iZipLen != 5 && iZipLen != 9 ){
		this.error = "The " + description + " field must contain either 5 or 9 digits.";
	}
}
_addValidator("isZipCode", _Field_isZipCode);

// define Field isFormat(); prototype
function _Field_isFormat(mask, type){
	var mask = _param(arguments[0]);
	var type = _param(arguments[1], "numeric").toLowerCase();
	var strErrorMsg = "";

	var strMaskLC = mask.toLowerCase();
	// define quick masks
	if( strMaskLC == "ssn" ){
		mask = "xxx-xx-xxxx";
		type = "numeric";
		var description = (this.description == this.name.toLowerCase()) ? "social security number" : this.description;
		strErrorMsg = "The " + description + " field must contain 9 digits and \nshould be in the format: " + mask;

	} else if( (strMaskLC == "phone") || (strMaskLC == "phone1") ){
		mask = "(xxx) xxx-xxxx";
		type = "numeric";
		var description = (this.description == this.name.toLowerCase()) ? "phone number" : this.description;
		strErrorMsg = "The " + description + " field must contain 10 digits and \nshould be in the format: " + mask;

	} else if( strMaskLC == "phone2" ){
		mask = "xxx-xxx-xxxx";
		type = "numeric";
		var description = (this.description == this.name.toLowerCase()) ? "phone number" : this.description;
		strErrorMsg = "The " + description + " field must contain 10 digits and \nshould be in the format: " + mask;

	} else if( strMaskLC == "phone3" ){
		mask = "xxx/xxx-xxxx";
		type = "numeric";
		var description = (this.description == this.name.toLowerCase()) ? "phone number" : this.description;
		strErrorMsg = "The " + description + " field must contain 10 digits and \nshould be in the format: " + mask;

	} else if( strMaskLC == "phone7" ){
		mask = "xxx-xxxx";
		type = "numeric";
		var description = (this.description == this.name.toLowerCase()) ? "phone number" : this.description;
		strErrorMsg = "The " + description + " field must contain 7 digits and \nshould be in the format: " + mask;

	} else if( strMaskLC == "zip" ){
		if( _stripInvalidChars(this.value).length < 6 ){
			mask = "xxxxx";
		} else {
			mask = "xxxxx-xxxx";
		}
		type = "numeric";
		var description = (this.description == this.name.toLowerCase()) ? "zip code" : this.description;
		strErrorMsg = "The " + description + " field should contain either 5 or 9 digits \nand be in the format: xxxxx or xxxxx-xxxx";

	} else if( strMaskLC == "zip5" ){
		mask = "xxxxx";
		type = "numeric";
		var description = (this.description == this.name.toLowerCase()) ? "zip code" : this.description;
		strErrorMsg = "The " + description + " field should contain 5 digits \nand be in the format: " + mask;

	} else if( strMaskLC == "zip9" ){
		mask = "xxxxx-xxxx";
		type = "numeric";
		var description = (this.description == this.name.toLowerCase()) ? "zip code" : this.description;
		strErrorMsg = "The " + description + " field should contain 9 digits \nand be in the format: " + mask;
	} else {
		var description = this.description;
	}

	var string = _stripInvalidChars(this.value, type);
	var masklen = _stripInvalidChars(mask, "x").length;

	// check to make sure the string contains the correct number of characters
	if( string.length != masklen && this.value.length > 0){
		if( strErrorMsg.length == 0 ) strErrorMsg = "This field requires at least " + masklen + " valid characters. Please \nmake sure to enter the value in the format: \n   " + mask + "\n(where 'x' is a valid character.)";
		this.error = strErrorMsg;

	// else re-format the string as defined by the mask
	} else if( string.length == masklen ){
		// find the position of all non "X" characters
		var stcMask = new Object();
		var lc = mask.toLowerCase();
		// loop through the string an make sure each character is an valid character
		for( var i=0; i < mask.length; i++ ){
			if( lc.charAt(i) != "x" ) stcMask[i] = mask.charAt(i);
	  }

		// put all the non-"X" characters back into the parsed string
		var iLastChar = 0;
		var newString = "";
		var i = 0;
		for( var pos in stcMask ){
			pos = parseInt(pos, 10);
			if( pos > iLastChar ){
				newString += string.substring(iLastChar, pos-i) + stcMask[pos];
				iLastChar = pos-i;
			} else {
				newString += stcMask[pos];
			}
			i++;
		}
		if( i == 0 ){
			newString = string;
		} else {
			newString += string.substring(iLastChar);
		}

		// set the value of the field to the new string--make sure not to perform the onBlur event
		this.setValue(newString, true, false);
	}
}
_addValidator("isFormat", _Field_isFormat);

// define Field isLengthGT(); prototype
function _Field_isLengthGT(len){
	if( this.obj.value.length <= len){
		this.error = "The " + this.description + " field must be greater than " + len + " characters.";
	}
}
_addValidator("isLengthGT", _Field_isLengthGT);

// define Field isLengthLT(); prototype
function _Field_isLengthLT(len){
	if( this.obj.value.length >= len){
		this.error = "The " + this.description + " field must be less than " + len + " characters.";
	}
}
_addValidator("isLengthLT", _Field_isLengthLT);



trunk/js/lib/qforms/functions_js12.js
/******************************************************************************
 qForm JSAPI: Functions Library

 Author: Dan G. Switzer, II
 Build:  111
******************************************************************************/
function _trim(s){
	return _rtrim(_ltrim(s));
}

function _ltrim(s){
	// remove all beginning white space
	return (s.length == 0) ? s : s.replace(new RegExp("^\\s+", qFormAPI.reAttribs), "");
}

function _rtrim(s){
	// remove all ending white space
	return (s.length == 0) ? s : s.replace(new RegExp("\\s+$", qFormAPI.reAttribs), "");
}

function _listToArray(string,delim){
	var delim = _param(arguments[1], ",");
	tmp = string.split(delim);
	for( var i=0; i < tmp.length; i++ ) tmp[i] = _trim(tmp[i]);
	return tmp;
}

function _listSum(string,delim){
	var delim = _param(arguments[1], ",");
	tmp = _listToArray(string,delim);
	iValue = 0;
	for( var i=0; i < tmp.length; i++ ) iValue += parseInt(_trim(tmp[i]), 10);
	return iValue;
}

function _stripInvalidChars(_s, _t){
	var s = _param(arguments[0]);
	var t = _param(arguments[1], "numeric").toLowerCase();
	var r;

	if( t == "numeric" ) r = new RegExp("[^0-9]+", qFormAPI.reAttribs);
	else if( t == "alpha" ) r = new RegExp("[^A-Za-z]+", qFormAPI.reAttribs);
	else if( t == "alphanumeric" ) r = new RegExp("\\W+", qFormAPI.reAttribs);
	else r = new RegExp("[^" + t + "]+", qFormAPI.reAttribs);

	s = s.replace(r, "");

	return s;
}

function _isLength(string, len, type){
	var string = _param(arguments[0]);
	var len = parseInt(_param(arguments[1], 10, "number"), 10);
	var type = _param(arguments[2], "numeric");

	var tmp = _stripInvalidChars(string, type);
	return (tmp.length == len) ? true : false;
}

function _getState(abbr){
	var abbr = _param(arguments[0]).toLowerCase();
	_s = new Object(); _s.al = "Alabama"; _s.ak = "Alaska"; _s.as = "American Samoa"; _s.az = "Arizona"; _s.ar = "Arkansas"; _s.ca = "California"; _s.co = "Colorado"; _s.ct = "Connecticut"; _s.de = "Delaware"; _s.dc = "District of Columbia"; _s.fm = "Federal States of Micronesia"; _s.fl = "Florida"; _s.ga = "Georgia"; _s.gu = "Guam"; _s.hi = "Hawaii"; _s.id = "Idaho"; _s.il = "Illinois"; _s["in"] = "Indiana"; _s.ia = "Iowa"; _s.ks = "Kansas"; _s.ky = "Kentucky"; _s.la = "Louisana"; _s.me = "Maine"; _s.mh = "Marshall Islands"; _s.md = "Maryland"; _s.ma = "Massachusetts"; _s.mi = "Michigan"; _s.mn = "Minnesota"; _s.ms = "Mississippi"; _s.mo = "Missouri"; _s.mt = "Montana"; _s.ne = "Nebraska"; _s.nv = "Nevada"; _s.nh = "New Hampshire"; _s.nj = "New Jersey"; _s.nm = "New Mexico"; _s.ny = "New York"; _s.nc = "North Carolina"; _s.nd = "North Dakota"; _s.mp = "Northern Mariana Islands"; _s.oh = "Ohio"; _s.ok = "Oklahoma"; _s.or = "Oregon"; _s.pw = "Palau"; _s.pa = "Pennsylvania"; _s.pr = "Puerto Rico"; _s.ri = "Rhode Island"; _s.sc = "South Carolina"; _s.sd = "South Dakota"; _s.tn = "Tennessee"; _s.tx = "Texas"; _s.ut = "Utah"; _s.vt = "Vermont"; _s.vi = "Virgin Islands"; _s.va = "Virginia"; _s.wa = "Washington"; _s.wv = "West Virginia"; _s.wi = "Wisconsin"; _s.wy = "Wyoming"; _s.aa = "Armed Forces Americas"; _s.ae = "Armed Forces Africa/Europe/Middle East"; _s.ap = "Armed Forces Pacific";
	if( !_s[abbr] ){
		return null;
	} else {
		return _s[abbr];
	}
}

// define the default properties for the sort
qFormAPI.sortOptions = new Object();
qFormAPI.sortOptions.order = "asc";
qFormAPI.sortOptions.byText = true;

function _sortOptions(obj, order, byText){
	var order = _param(arguments[1], qFormAPI.sortOptions.order);
	if( order != "asc" && order != "desc" ) order = "asc";
	var byText = _param(arguments[2], qFormAPI.sortOptions.byText, "boolean");
	var orderAsc = (order == "asc") ? true : false;

	// loop through all the options and sort them asc
	for( var i=0; i < obj.options.length; i++ ){
		for( var j=0; j < obj.options.length-1; j++ ){
			// if an option is greater than the next option, swap them
			if( orderAsc && (byText && obj.options[j].text > obj.options[j+1].text) || (!byText && obj.options[j].value > obj.options[j+1].value) ){
				_swapOptions(obj.options[j], obj.options[j+1]);
			} else if( !orderAsc && (byText && obj.options[j].text < obj.options[j+1].text) || (!byText && obj.options[j].value < obj.options[j+1].value) ){
				_swapOptions(obj.options[j], obj.options[j+1]);
			}
		}
	}
	return true;
}

function _swapOptions(o1, o2){
	var sText = o1.text;
	var sValue = o1.value;
	var sSelected = o1.selected;
	o1.text = o2.text ;
	o1.value = o2.value;
	o1.selected = o2.selected;
	o2.text = sText;
	o2.value = sValue;
	o2.selected = sSelected;
}

function _transferOptions(field1, field2, sort, type, selectItems, reset){
	var sort = _param(arguments[2], true, "boolean");
	var type = _param(arguments[3], "selected").toLowerCase();
	if( type != "all" && type != "selected" ) type = "selected";
	var selectItems = _param(arguments[4], true, "boolean");
	var reset = _param(arguments[5], false, "boolean");
	var doAll = (type == "all") ? true : false;

	if( field1.type.substring(0,6) != "select" ) return alert("This method is only available to select boxes. \nThe field \"" + field1.name + "\" is not a select box.");
	if( field2.type.substring(0,6) != "select" ) return alert("This method is only available to select boxes. \nThe field \"" + field2.name + "\" is not a select box.");

	// clear the select box
	if( reset ) field2.length = 0;

	for( var i=0; i < field1.length; i++ ){
		// if the current option is selected, move it
		if( doAll || field1.options[i].selected ){
			field2.options[field2.length] = new Option(field1.options[i].text, field1.options[i].value, false, selectItems);
			field1.options[i] = null;
			i--; // since you just deleted a option, redo this array position next loop
		}
	}

	// if sorting the fields
	if( sort ) _sortOptions(field2);
	return true;
}


function _getURLParams(){
	struct = new Object();
	var strURL = document.location.href;
	var iPOS = strURL.indexOf("?");
	// if there are some query string params, split them into an array
	if( iPOS != -1 ){
		var strQS = strURL.substring(iPOS + 1);
		var aryQS = strQS.split("&");
	// otherwise, return the empty structure
	} else {
	 return struct;
	}

	// loop through the array
	for( var i=0; i < aryQS.length; i++ ){
		iPOS = aryQS[i].indexOf("=");
		// if no equal sign is found, then the value is null
		if( iPOS == -1 ){
			struct[aryQS[i]] = null;
		// otherwise grab the variable name and it's value and stick it in structure
		} else {
			var key = aryQS[i].substring(0, iPOS);
			var value = unescape(aryQS[i].substring(iPOS+1));
			// if the value doesn't exist, then create a new key
			if( !struct[key] ) struct[key] = value;
			// otherwise, append the value
			else struct[key] += "," + value;
		}
	}
	return struct;
}

function _createFields(struct, type){
	var type = _param(arguments[1], "hidden");
	if( this.status == null ) return false;
	// loop through form elements
	for( key in struct ){
		document.write("<input type=\"" + type + "\" name=\"" + key + "\" value=\"" + struct[key] + "\" />");
	}
	return true;
}

function _getEventType(type){
	// the default event type
	var strEvent = "onblur";
	// if this is a checkbox & radio button, then mirror value on click
	if( type == "checkbox" || type == "radio" ) strEvent = "onclick";
	// if this is a select box, then mirror value when the value changes
	else if( type.substring(0,6) == "select" ) strEvent = "onchange";
	return strEvent;
}




MachIIDashboard/assets/js/generic.js
/*
 * License:
 * Copyright 2008 GreatBizTools, LLC
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * 
 * Copyright: GreatBizTools, LLC
 * Author: Peter J. Farrell (peter@mach-ii.com)
 * $Id: generic.js 1038 2008-09-14 07:10:28Z peterfarrell $
 * 
 */
function addLoadEvent(func) {
  var oldonload = window.onload;
  if (typeof window.onload != 'function') {
    window.onload = func;
  } else {
    window.onload = function() {
      if (oldonload) {
        oldonload();
  }
      func();
    }
  }
} // END FUNCTION

function fancyRules() { 
  if (!document.getElementsByTagName) return; 
	  var hr = document.getElementsByTagName("hr");
  for (var i=0; i<hr.length; i++) { 
    var newhr = hr[i]; 
    var wrapdiv = document.createElement('div');
    wrapdiv.className = 'line';  
    newhr.parentNode.replaceChild(wrapdiv, newhr);  
    wrapdiv.appendChild(newhr);  
  }
} 